_posts/2020/2020-02-02-calculate-a-mean-average-from-a-ruby-array.md

```diff
@@ -0,0 +1,130 @@
+
+In order to see the speed of the different implementations we can use the `benchmark-ips` gem, which compares the performance of different bits of code.
+
+```ruby
+require "benchmark/ips"
+
+# Generate a 1,000 element array
+a = Array.new(1000) { |_| rand(1000) }
+
+Benchmark.ips do |x|
+  x.report("sum(0.0) / size") do
+    a.sum(0.0) / a.size
+  end
+  x.report('inject(0.0) / size') do
+    a.inject(0.0) { |result, i| result + i } / a.size
+  end
+  x.compare!
+end
+```
+
+The results (on my laptop) show a 50× speed improvement when using the native `#sum` in the implementation. This is due to it being written into the language to provide exactly this sort of performance improvement.
+
+```
+Calculating -------------------------------------
+   sum(0.0) / size  680.425k (± 6.5%) i/s -  3.432M in 5.06s
+inject(0.0) / size   13.513k (± 5.0%) i/s - 67.586k in 5.01s
+
+Comparison:
+   sum(0.0) / size: 680425.2 i/s
+inject(0.0) / size:  13512.7 i/s - 50.35x  slower
+```
+
+Although there are variations using `#sum` and then `#size` that perform similarly, the most important difference is clearly the use of the native `#sum` method.
+
+The benchmarking code for a wider variety of implementations...
+
+```ruby
+require "benchmark/ips"
+
+# Generate a 1,000 element array
+a = Array.new(1000) { |_| rand(1000) }
+
+Benchmark.ips do |x|
+  x.report("sum(0.0) / size") do
+    a.sum(0.0) / a.size
+  end
+  x.report("sum.to_f / size") do
+    a.sum.to_f / a.size
+  end
+  x.report("sum / size.to_f") do
+    a.sum / a.size.to_f
+  end
+  x.report("sum.fdiv(size)") do
+    a.sum.fdiv(a.size)
+  end
+  x.report('inject(0.0, :+) / size') do
+    a.inject(0.0, :+) / a.size
+  end
+  x.report('inject(0.0) / size') do
+    a.inject(0.0) { |result, i| result + i } / a.size
+  end
+  x.report('inject(0).to_f / size') do
+    a.inject(0) { |result, i| result + i }.to_f / a.size
+  end
+  x.report('inject(0) / size.to_f') do
+    a.inject(0) { |result, i| result + i } / a.size.to_f
+  end
+  x.report('inject(0).fdiv(size)') do
+    a.inject(0) { |result, i| result + i }.fdiv(a.size)
+  end
+  x.compare!
+end
+```
+
+...and the comparison results:
+
+```
+Comparison:
+       sum(0.0) / size: 668222.4 i/s
+       sum / size.to_f: 660291.4 i/s - same-ish
+       sum.to_f / size: 655929.1 i/s - same-ish
+        sum.fdiv(size): 621960.0 i/s - same-ish
+inject(0.0, :+) / size:  30823.6 i/s - 21.68x slower
+ inject(0) / size.to_f:  18740.7 i/s - 35.66x slower
+ inject(0).to_f / size:  18320.1 i/s - 36.47x slower
+  inject(0).fdiv(size):  18082.6 i/s - 36.95x slower
+    inject(0.0) / size:  15264.0 i/s - 43.78x slower
+```
+
+
+## Anything else?
+
+Rails’ Active Record does include an [`#average` method](https://api.rubyonrails.org/classes/ActiveRecord/Calculations.html#method-i-average). It’s used to perform calculations on numerical columns in the database directly in SQL.
```

Comment:
Technically this should be "Rails's" even though, for a long time, stylistically, I'd write the way you did: https://grammar.yourdictionary.com/punctuation/apostrophe-rules.html